### ISSUE TITLE
Deserialize multiple cache buffers to one FiltersEngine instance

### ISSUE THREAD

**Issue by sentialx** [cliqz-oss/adblocker#132]
Hi, I have one cache file with easylist, easyprivacy etc. and I would like to also include optional, regional filters, but it seems there's no other way to do that without creating multiple FiltersEngine instances. Is there a way to do it like this?
```javascript
const engine = new FiltersEngine();
engine.deserialize(buffer1);
engine.deserialize(buffer2);
```
I've also seen the update method, but why distinguishing cosmetic filters from network when the  `FiltersEngine.parse` method does that, but it can be used only once per instance?

**Comment by remusao** [comment]
Hi @sentialx,

Thanks for taking the time to open an issue! What you are trying to do it currently not supported but could be in the future. We did not have this use-case on our side so far. I was thinking in the past on how we could implement this merging of serialized structures efficiently; I believe it should be possible and probably not too hard. It remains to be seen how much faster it would be compared to just initializing the engine again or calling `update`, since internal structures would need to be allocated and initialized anyway. So for now I believe there is no way around using the `update` method from `FiltersEngine` to add filters.

> why distinguishing cosmetic filters from network when the FiltersEngine.parse method does that

Currently we use the `FiltersEngine.update` method alongside the `parseFilters` helper which is also exposed by the library (in fact, it's what the `FiltersEngine.parse` static method is doing internally). It gets a string to parse (e.g.: Easylist or Easyprivacy) and returns both cosmetic and network filters separated; which you can directly forward to `FiltersEngine.update` (https://github.com/cliqz-oss/adblocker/blob/master/src/lists.ts#L90). We could add a helper method on `FiltersEngine` to abstract this away.

I have a few things I'd like to finish before but I will think about implementing this merging soon, I agree that this would be a nice addition to the API.

Best,
Rémi

**Comment by sentialx** [comment]
Thanks for your fast response. The helper function to separate the filter lists to cosmetic and network would be really helpful.

**Comment by remusao** [comment]
No worries! In the meanwhile this could be done in the following way (until we implement the helper in `FiltersEngine` directly):

```javascript
const { FiltersEngine, parseFilters } = require('@cliqz/adblocker');

const engine = FiltersEngine.parse('||foo.com');
const { networkFilters, cosmeticFilters } = parseFilters('||bar.com', engine.config);
engine.update({
  newNetworkFilters: networkFilters,
  newCosmeticFilters: cosmeticFilters,
});
```

**Comment by sentialx** [comment]
Oh, so you have a nice way to parse the filters. Perhaps I missed something

**Comment by sentialx** [comment]
Can the parseFilters method parse multiple filters or only one?

**Comment by remusao** [comment]
Yes you can give it a string of multiple lines. In fact it's totally fine to feed it the raw content of easylist.txt for example (or any other for that matter).

```js
const { parseFilters } = require('@cliqz/adblocker');
const { networkFilters, cosmeticFilters } = parseFilters(`
[Adblock Plus 2.0]
! Version: 201904171930
! Title: EasyList
! Last modified: 17 Apr 2019 19:30 UTC
! Expires: 4 days (update frequency)
! Homepage: https://easylist.to/
! Licence: https://easylist.to/pages/licence.html
! 
! Please report any unblocked adverts or problems
! in the forums (https://forums.lanik.us/)
! or via e-mail (easylist.subscription@gmail.com).
! GitHub issues: https://github.com/easylist/easylist/issues
! GitHub pull requests: https://github.com/easylist/easylist/pulls
! 
! -----------------------General advert blocking filters-----------------------!
! *** easylist:easylist/easylist_general_block.txt ***
&act=ads_
&ad.vid=$~xmlhttprequest
&ad_box_
&ad_channel=
&ad_classid=
&ad_code=
&ad_height=
&ad_ids=
&ad_keyword=
&ad_network_
&ad_number=
&ad_revenue=
&ad_slot=
&ad_time=
&ad_type=
...
`);
```

You can invoke it N times to parse your N lists (easylist, easyprivacy, etc.) or just concatenate all of the lists into one big string and parse it in one go. Up to you!

**Comment by sentialx** [comment]
Nice! Thanks for your help!

**Comment by remusao** [comment]
Closing this as resolved. Feel free to re-open or open another issue if I can help in any way.
